controller/lib/core/flow_integrator.cpp

```diff
@@ -14,63 +14,78 @@ limitations under the License.
-// margin.
-//
-// Even 40ml/s may seem way higher than should ever occur, but it's actually
-// easy to get an error this high at zero flow.  Since we have two sensors,
-// we'd need an error of +/-20ml in each.  At zero flow, this corresponds to
-// significantly less than 1 Pa of pressure, and just two or three bits in our
-// 12-bit ADC!  See
-// https://docs.google.com/spreadsheets/d/1G9Kb-ImlluK8MOx-ce2rlHUBnTOtAFQvKjjs1bEhlpM
-//
-// In theory the story is better at higher flows (because the venturi response
-// is the square root of pressure), and we may run a bias flow through our
-// device to put us into a regime where we're less sensitive to error
-// (equivalently, where the venturi response has a lower slope).  But we still
-// want to handle the low-flow case as best we can.
-static constexpr float FLOW_PID_KU = 0.20f;
-static constexpr float FLOW_PID_TU = 5;
-static constexpr VolumetricFlow FLOW_PID_OUTPUT_MAX = ml_per_sec(100);
-
-// TODO: Make PID parameters configurable.
-
-FlowIntegrator::FlowIntegrator()
-    : flow_correction_pid_(
-          /*kp=*/0.6f * FLOW_PID_KU,
-          /*ki=*/1.2f * FLOW_PID_KU / FLOW_PID_TU,
-          /*kd=*/3 * FLOW_PID_KU * FLOW_PID_TU / 40,
-          // ProportionalTerm::ON_ERROR and DifferentialTerm::ON_MEASUREMENT
-          // makes for your "standard" PID.  TODO: perhaps these shouldn't even
-          // be options.
-          ProportionalTerm::ON_ERROR,        //
-          DifferentialTerm::ON_MEASUREMENT,  //
-          -FLOW_PID_OUTPUT_MAX.ml_per_sec(), //
-          FLOW_PID_OUTPUT_MAX.ml_per_sec()) {}
+FlowIntegrator::FlowIntegrator() = default;
 
 void FlowIntegrator::AddFlow(Time now, VolumetricFlow uncorrected_flow) {
-  VolumetricFlow flow = uncorrected_flow + flow_correction_;
-  Duration delta = now - last_flow_measurement_time_;
-  if (delta >= VOLUME_INTEGRAL_INTERVAL) {
-    volume_ += delta * (last_flow_ + flow) / 2.0f;
+  VolumetricFlow flow = uncorrected_flow + FlowCorrection();
+
+  // First time?  In that case, don't need to adjust the flow.  Just record
+  // this data point.
+  if (last_flow_measurement_time_ == std::nullopt) {
+    last_note_time_ = now;
+    last_note_vol_ = ml(0);
+    last_flow_measurement_time_ = now;
+    last_flow_ = flow;
+    return;
+  }
+
+  Duration dt = now - *last_flow_measurement_time_;
+  if (dt >= VOLUME_INTEGRAL_INTERVAL) {
+    volume_ += dt * (last_flow_ + flow) / 2.0f;
     last_flow_measurement_time_ = now;
     last_flow_ = flow;
   }
 }
 
 void FlowIntegrator::NoteExpectedVolume(Volume v) {
-  flow_correction_ = ml_per_sec(flow_correction_pid_.Compute(
-      last_flow_measurement_time_, volume_.ml(), v.ml()));
+  // We need at least one call to AddFlow before we can do anything.
+  if (last_note_time_ == std::nullopt ||
+      last_flow_measurement_time_ == std::nullopt ||
+      *last_note_time_ >= *last_flow_measurement_time_) {
+    return;
+  }
+
```

Comment:
I'd add a couple sentences to describe the spirit of this. Eg:

The code below talks about 3 moments: t0 (time of the previous breath / NoteExpectedVolume), t1 (the current time) and t2 (projected time of the next breath). We assume that they are equally spaced by "dt" (This is wrong if breaths are of different length, but most of the time it is true).

We assume that the uncorrected volume drifts due to sensor inaccuracies by approximately "dv" over each breath. [question: shouldn't we then be computing flow drift instead of volume drift? That would be more resilient to different breath lengths].

Finally, we compute a flow correction that will compensate for dv over the next breath, making corrected volume at "t2" hopefully be very close to the current corrected volume at "t1".

Reply:
> [question: shouldn't we then be computing flow drift instead of volume drift? That would be more resilient to different breath lengths].

Yeah, this gets to the question of, there are actually two corrections I'm making here.  One is this constant correction: I see volume is X != 0, I want to drive that to 0 over the upcoming breath.  The other is flow rate correction: I see that volume is increasing at rate Y != 0, and I want to correct for that.

I'm combining these two, and that's not really correct.

> making corrected volume at "t2" hopefully be very close to the current corrected volume at "t1".

Importantly, that's not what I'm doing.  I want corrected volume at t2 to hopefully be very close to the given desired volume `v`.